src/util/group_pool.rs

```diff
@@ -0,0 +1,569 @@
+/// is ready to process a task, it get a task from the waiting queue
+/// according the following rules:
+///
+/// ## Choose one group
+/// 1. Choose the groups with least running tasks.
+/// 2. If more than one groups has the least running tasks,
+///    choose the one whose task comes first(with the minum task's ID)
+///
+/// ## Chose a task in selected group.
+/// For the tasks in the selected group, choose the first
+/// one in the queue(which means comes first).
+///
+/// # Example
+///
+/// One group with 10 tasks, and 10 groups with 1 task each. Every
+/// task cost the same time.
+///
+/// ```
+/// # extern crate tikv;
+/// use tikv::util::group_pool::GroupTaskPool;
+/// use std::thread::sleep;
+/// use std::time::Duration;
+/// use std::sync::mpsc::{Sender, Receiver, channel};
+/// # fn main(){
+/// let concurrency = 2;
+/// let name = Some(String::from("sample"));
+/// let mut task_pool = GroupTaskPool::new(name,concurrency);
+/// let (jtx, jrx): (Sender<u64>, Receiver<u64>) = channel();
+/// let group_with_many_tasks = 1001 as u64;
+/// // all tasks cost the same time.
+/// let sleep_duration = Duration::from_millis(100);
+///
+/// // push tasks of group_with_many_tasks's job to make all thread in pool busy.
+/// // in order to make sure the thread would be free in sequence,
+/// // these tasks should run in sequence.
+/// for _ in 0..concurrency {
+///     task_pool.execute(group_with_many_tasks, move || {
+///         sleep(sleep_duration);
+///     });
+///     // make sure the pre task is running now.
+///     sleep(sleep_duration / 4);
+/// }
+///
+/// // push 10 tasks of group_with_many_tasks's job into pool.
+/// for _ in 0..10 {
+///     let sender = jtx.clone();
+///     task_pool.execute(group_with_many_tasks, move || {
+///         sleep(sleep_duration);
+///         sender.send(group_with_many_tasks).unwrap();
+///     });
+/// }
+///
+/// // push 1 task for each group_id in [0..10) into pool.
+/// for group_id in 0..10 {
+///     let sender = jtx.clone();
+///     task_pool.execute(group_id, move || {
+///         sleep(sleep_duration);
+///         sender.send(group_id).unwrap();
+///     });
+/// }
+/// // when first thread is free, since there would be another
+/// // thread running group_with_many_tasks's task,and the first task which
+/// // is not belong to group_with_many_tasks in the waitting queue would run first.
+/// // when the second thread is free, since there is group_with_many_tasks's task
+/// // is running, and the first task in the waiting queue is the
+/// // group_with_many_tasks's task,so it would run. Similarly when the first is
+/// // free again,it would run the task in front of the queue..
+/// for id in 0..10 {
+///     let first = jrx.recv().unwrap();
+///     assert_eq!(first, id);
+///     let second = jrx.recv().unwrap();
+///     assert_eq!(second, group_with_many_tasks);
+/// }
+/// # }
+/// ```
+pub struct GroupTaskPool {
+    tasks: Arc<Mutex<GroupTaskPoolMeta>>,
```

Comment:
Since all threads share one `GroupTaskPoolMeta ` to get the next task to process. @BusyJay